sds/src/scoped_ruleset/mod.rs

```diff
@@ -144,23 +192,28 @@ where
 }
 
 pub struct RuleIndexVisitor<'a> {
-    tree_nodes: &'a Vec<&'a RuleTree>,
+    tree_nodes: &'a Vec<ActiveRuleTree<'a>>,
+    used_rule_set: &'a mut BoolSet,
 }
 
 impl<'a> RuleIndexVisitor<'a> {
     /// Visits all rules associated with the current string. This may
     /// potentially return no rule indices at all.
-    pub fn visit_rule_indices(&self, mut visit: impl FnMut(usize)) {
+    pub fn visit_rule_indices(&mut self, mut visit: impl FnMut(usize)) {
         // visit rules with an `Include` scope
         for include_node in self.tree_nodes {
-            for change in &include_node.rule_changes {
+            if include_node.index_wildcard_match {
+                // This is guaranteed to be a duplicated node. Skip it
+                continue;
+            }
+            for change in &include_node.rule_tree.rule_changes {
                 match change {
                     RuleChange::Add(rule_id) => {
-                        (visit)(*rule_id);
-                    }
-                    RuleChange::Remove(_) => {
-                        /* ignore removals, they can be checked as needed later */
+                        if !self.used_rule_set.get_and_set(*rule_id) {
+                            (visit)(*rule_id);
```

Comment:
I think it's rule_index here (not anymore a rule id)


```suggestion
                    RuleChange::Add(rule_index) => {
/* ignore removals, they can be checked as needed later */
                        if !self.used_rule_set.get_and_set(*rule_index) {
                            (visit)(*rule_index);